backends/ref/ceed-ref-operator.c

```diff
@@ -664,11 +664,69 @@ static inline void CeedOperatorGetBasisPointer_Ref(const CeedScalar **basisptr,
   }
 }
 
+//------------------------------------------------------------------------------
+// Create CSR restriction
+//------------------------------------------------------------------------------
+static int CreateCSRRestriction(CeedElemRestriction rstr,
+                                CeedElemRestriction *csrRstr) {
+  int ierr;
+
+  Ceed ceed;
+  ierr = CeedElemRestrictionGetCeed(rstr, &ceed); CeedChk(ierr);
+  const CeedInt *offsets;
+  ierr = CeedElemRestrictionGetOffsets(rstr, CEED_MEM_HOST, &offsets);
+  CeedChk(ierr);
+  CeedVector mult;
+  ierr = CeedElemRestrictionCreateVector(rstr, &mult, NULL); CeedChk(ierr);
+  ierr = CeedVectorSetValue(mult, 0.0); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetMultiplicity(rstr, mult); CeedChk(ierr);
+
+  // Account for gaps in offsets
+  CeedInt length, shift = 0;
+  CeedScalar *multArray;
+  ierr = CeedVectorGetLength(mult, &length); CeedChk(ierr);
+  ierr = CeedVectorGetArray(mult, CEED_MEM_HOST, &multArray); CeedChk(ierr);
+  for (CeedInt i = 0; i < length; i++) {
+    if (fabs(multArray[i]) < CEED_EPSILON*10)
+      shift++;
+    multArray[i] = shift;
+  }
+
+  // Compact offsets
+  CeedInt nelem, ncomp, elemsize, compstride, max = 1, *csrOffsets;
+  ierr = CeedElemRestrictionGetNumElements(rstr, &nelem); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetNumComponents(rstr, &ncomp); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetElementSize(rstr, &elemsize); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetCompStride(rstr, &compstride); CeedChk(ierr);
+  shift = ncomp;
+  if (compstride != 1)
+    shift *= ncomp;
+  ierr = CeedCalloc(nelem*elemsize, &csrOffsets); CeedChk(ierr);
+  for (CeedInt i = 0; i < nelem*elemsize; i++) {
+    csrOffsets[i] = (offsets[i] - (CeedInt)multArray[offsets[i]])*shift;
+    if (csrOffsets[i] > max)
+      max = csrOffsets[i];
+  }
```

Comment:
The composite operator case for mixed finite elements is why I'm filling in the gaps.

Consider the future mixed finite elements code for the solids miniapp. We will want to assemble the point block diagonal for just the pressure fields. But the pressure only sub-operator will have no knowledge of the displacement fields. For that operator, the input vector will look like
```
[ ????????? E0P0, E0Px, E0Py, E0Pz, ?????? E1P0, E1Px, E1Py, E1Pz, ??????]
```
with no information about the portions of the input vector that do not correspond to the pressure DoFs. I need to make a new restriction from the old restriction for `4*4` components of the point block diagonal instead of `4` components, but then what to do with the gaps? This sub-operator has no idea that the gaps represent three components of displacement.

Reply:
So in the case of surface integrals, we generally want to sum all the sub-operators to get a global pbdiagonal. The Schur complement of displacement in pressure needs a special weighted inner product so it kinda needs to be a new operator anyway. When PCFieldSplit manipulates the Schur complement, there is a T-vector containing only pressure. That has no overlap (because pressure is discontinuous), so it's the same data as the L-vector. You'd have a separate restriction from that contiguous L-vector.

If we were to ask for a pbdiagonal of a mixed FE operator, I would do it by enumerating the nodes of the L-vector, squaring their size, and scanning to get new offsets.

Reply:
So in practice we won't have gaps unless we have a more complex operator that this code currently isn't designed to handle?

Reply:
I think this is all set for single field now. I'm sketching out how I would add composite operator support, and I think it should be pretty straightforward. Any preference between adding to this PR of a separate one?